0002-crypto-conditions/0002-crypto-conditions.md

```diff
@@ -114,312 +114,288 @@ This specification is a part of the [Interledger Protocol](https://interledger.o
+1. The implementation must derive a condition from the fulfillment and ensure that the derived condition (D) matches the given condition (C). 
 
-### Example Condition
+2. Any signatures in the fulfillment (F) must be verified, using the appropriate signature verification algorithm, against the corresponding public key, also provided in the fulfillment and the message (M), which may be empty.
 
-An example condition in string format:
+If the derived condition (D) matches the input condition (C) AND the boolean circuit defined by the fulfillment evaluates to TURE then the fullfillment (F) fulfills the condition (C).
 
-    cc:0:3:dB-8fb14MdO75Brp_Pvh4d7ganckilrRl13RS_UmrXA:66
+## Conditions as Public Keys
 
-The example has the following attributes:
+Since the condition is just a fingerprint and meta-data about the crypto-condition it can be transmitted freely in the same way a public key is shared publicly. It's not possible to derive the fulfillment from the condition.
 
-| Field                | Value | Description |
-|----------------------|-------|----------------------------------------------|
-| preface              | `cc`  | Constant. Indicates this is a condition. |
-| type                 | `0`   | Type 0 is [PREIMAGE-SHA-256][]. |
-| featuresBitmask      | `3`   | Boolean-OR combination of feature suites SHA-256 (feature bit 0x01) and PREIMAGE (feature bit 0x02). |
-| fingerprint          | `dB-8fb14MdO75Brp_Pvh4d7ganckilrRl13RS_UmrXA` | The hash of the fulfillment for this condition. |
-| maxFulfillmentLength | `66`  | The fulfillment payload is 66 bytes long, before being BASE64URL-encoded. |
-
-
-## Fulfillment {#fulfillment-format}
-Below are the string and binary encoding formats for a fulfillment.
-
-### String Format {#string-fulfillment-format}
+# Format
 
-Fulfillments are ASCII encoded as:
+A description of crypto-conditions is provided in this document using Abstract Syntax Notation One (ASN.1) as defined in [ITU.X680.2015](#itu.X680.2015). Implementations of this spec MUST support encoding and decoding using Octet Encoding Rules (OER) as defined in [ITU.X696.2015](#itu.X696.2015).
 
-    "cf:" BASE16(type) ":" BASE64URL(payload)
+## Condition {#condition-format}
 
-### Binary Format {#binary-fulfillment-format}
+The binary encoding of conditions is dependant on their type. The overall size of conditions is not fixed as some types (compound conditions) contain additional fields (subtypes) and the algorithm for deriving the fingerprint differs by type therefor the fingerprint size also differs.
 
-Fulfillments are binary encoded as:
+Conditions are encoded as follows:
 
-    Fulfillment ::= SEQUENCE {
+    Condition ::= SEQUENCE {
       type ConditionType,
-      payload OCTET STRING
+      fingerprint OCTET STRING,
+      maxFulfillmentLength INTEGER (0..MAX),
+      subtypes OCTET STRING
```

Comment:
> I don't see how this is simpler than the features we currently have; in fact I think it's marginally worse, but I don't really feel strongly enough about this to "demand" one over the other.

What value do you get at as an implementer knowing that a compound condition you're processing has subconditions requiring you to support certain features. What you're trying to determine is whether you can evaluate the fulfillment of those subconditions later and that requires more than just the features it requires knowledge about those subcondition types too.

Consider a future where we add the SHA-512 feature suite and somebody implements support for RSA-SHA-256 and PREIMAGE-SHA-512 but not RSA-SHA-512 (bad example but not impossible).

They see a THRESHOLD-SHA-256 condition that says it has subconditions with the features SHA-512, SHA-256 and RSA. How do they know if that means they'll encounter an RSA-SHA-512 subfulfillment?

> First, since the field is, essentially, a bitstring, I'd suggest using the correct OER/ASN.1 type: BIT STRING.

I agree. @justmoon isn't crazy for them but given how things have evolved I think he may agree now too.

> Second, both the type and subtypes field should be BIT STRING. Rationale: they both contain the same kind of information: types; only difference is that one will only allow one flag to be set, while the other multiple flags.

This is an interesting idea and means we'll never run out of type IDs although it's obviously less efficient than a fixed size INTEGER. Again, I'm inclined to agree. We should use the right ASN.1 type and not think too much about the final encoded form. A few bytes here and there for length indicators is nothing relative to large prefixes and signatures.

> Last, I'd put the two fields next to each other.

Agreed. This is a hangover from trying to make it optional and make conditions constant size if you exclude the field. Not sure I see the value of that now.